File: frontend/src/db/schemas/order.ts
import type { RxJsonSchema } from 'rxdb'

export interface OrderItem {
  productId: string
  productName: string
  unitPrice: number
  quantity: number
  subtotal: number
}

export interface OrderDoc {
  id: string
  orderNumber: string
  items: OrderItem[]
  totalAmount: number
  paymentMethod: 'cash' | 'line_pay' | 'transfer'
  note: string | null
  deviceId: string
  createdAt: number
  syncedAt: number | null
}

export const orderSchema: RxJsonSchema<OrderDoc> = {
  version: 0,
  primaryKey: 'id',
  type: 'object',
  properties: {
    id: { type: 'string', maxLength: 36 },
    orderNumber: { type: 'string', maxLength: 50 },
    items: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          productId: { type: 'string' },
          productName: { type: 'string' },
          unitPrice: { type: 'number' },
          quantity: { type: 'number' },
          subtotal: { type: 'number' },
        },
        required: ['productId', 'productName', 'unitPrice', 'quantity', 'subtotal'],
      },
    },
    totalAmount: { type: 'number', multipleOf: 1, minimum: 0, maximum: 9999999 },
    paymentMethod: {
      type: 'string',
      enum: ['cash', 'line_pay', 'transfer'],
    },
    note: { type: ['string', 'null'] },
    deviceId: { type: 'string' },
    createdAt: { type: 'number', multipleOf: 1, minimum: 0, maximum: 99999999999999 },
    syncedAt: { type: ['number', 'null'] },
  },
  required: ['id', 'orderNumber', 'items', 'totalAmount', 'paymentMethod', 'deviceId', 'createdAt'],
  indexes: ['createdAt', 'orderNumber'],
}

File: backend/app/models/order.py
import uuid

from sqlalchemy import String, Integer, ForeignKey, BigInteger, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    exhibition_id: Mapped[str] = mapped_column(String(36), ForeignKey("exhibitions.id"))
    device_id: Mapped[str] = mapped_column(String(255))
    order_number: Mapped[str] = mapped_column(String(50), unique=True)
    total_amount: Mapped[int] = mapped_column(Integer)
    payment_method: Mapped[str] = mapped_column(String(20))
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[int] = mapped_column(BigInteger)
    cancelled_at: Mapped[int | None] = mapped_column(BigInteger, nullable=True)

    items: Mapped[list["OrderItem"]] = relationship(back_populates="order", cascade="all, delete-orphan")


class OrderItem(Base):
    __tablename__ = "order_items"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    order_id: Mapped[str] = mapped_column(String(36), ForeignKey("orders.id"))
    product_id: Mapped[str] = mapped_column(String(36))
    product_name: Mapped[str] = mapped_column(String(255))
    unit_price: Mapped[int] = mapped_column(Integer)
    quantity: Mapped[int] = mapped_column(Integer)
    subtotal: Mapped[int] = mapped_column(Integer)

    order: Mapped["Order"] = relationship(back_populates="items")
